### PR TITLE
MeasurementAgeUpdater test

### PR DESCRIPTION
**Pull request by DeanWay**

### CODE REVIEW

**Comment by DeanWay** [comment]
Nominating @JKereliuk and @mjshepherd for review

File: components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Before
+    public void setUp() throws NoSuchFieldException, IllegalAccessException
+    {
+        MockitoAnnotations.initMocks(this);
+        when(this.source.getXObject((EntityReference) Matchers.any())).thenReturn(this.patientRecordObj);
+
+        objects = new LinkedList<>();
+        objects.add(measurement);
+        when(this.source.getXObjects((EntityReference)Matchers.any())).thenReturn(objects);
```
Comment (mjshepherd) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: Matcher.<EntityReference>any()

File: components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        Date birthDate = dateFormat.parse(birthDateString);
+        Date measureDate = dateFormat.parse(measureDateString);
+
+        when(this.patientRecordObj.getDateValue(Matchers.matches("date_of_birth"))).thenReturn(birthDate);
+        when(this.measurement.getDateValue(DATE_PROPERTY_NAME)).thenReturn(measureDate);
+
+        this.mocker.getComponentUnderTest().onEvent(event, source, data);
+
+        verify(this.measurement).setFloatValue(eq(AGE_PROPERTY_NAME), eq(365/30.4375f));
+    }
```
Comment (mjshepherd) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: Maybe add tests for multiple measurement objects and no objects. Won't increase coverage but enforces expected behaviour

Comment (JKereliuk) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: To continue on this, do that and verify that certain methods are called multiple times with the correct arguments.

**Comment by mjshepherd** [comment]
More dates are good. But you should also check with multiple measurement objects. The MeasurementAgeUpdater retrieves all the measurement objects from the record and updates them all.

**Comment by DeanWay** [comment]
@mjshepherd Added tests for multiple and zero measurement objects

**Comment by mjshepherd** [comment]
+1

**Comment by JKereliuk** [comment]
+1 @sdumitriu your turn!

File: components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+ * Tests for the {@link MeasurementAgeUpdater} component.
+ *
+ */
+public class MeasurementAgeUpdaterTest
+{
+
+    @Rule
+    public MockitoComponentMockingRule<MeasurementAgeUpdater> mocker =
```
Comment (sdumitriu) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: The parameter should be the interface with the `@Role` annotation, in this case `EventListener`:

    public MockitoComponentMockingRule<EventListener> mocker =
        new MockitoComponentMockingRule<EventListener>(MeasurementAgeUpdater.class);

File: components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            new MockitoComponentMockingRule<>(MeasurementAgeUpdater.class);
```
Comment (sdumitriu) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: Our [codestyle](https://github.com/phenotips/build-tools/tree/master/codestyle-resources/src/main/resources) is to use one level of indentation. Although we don't enforce it with checkstyle for test code, it should still be respected as much as possible.

File: components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    @Test
+    public void returnsNormallyWhenGivenNoMeasurementObjects() throws Exception {
+        when(this.source.getXObjects(any(EntityReference.class))).thenReturn(null);
+        NullPointerException exceptionWhenListIsNull = null;
+        try {
+            this.mocker.getComponentUnderTest().onEvent(event, source, data);
+        } catch(NullPointerException e){
+            exceptionWhenListIsNull = e;
```
Comment (sdumitriu) [inline components/patient-measurements/api/src/test/java/org/phenotips/measurements/internal/MeasurementAgeUpdaterTest.java]: You don't need to do this. If there is an exception thrown, the test will automatically fail.

If, however, you do want to test that calling a method will throw an exception, then you would annotate the method with:

    @Test(expected=NullPointerException.class)

**Comment by sdumitriu** [comment]
You can install EclEmma to check which lines are covered by tests and which aren't. I think you're missing the case when the event's source doesn't have a patient object at all.

**Comment by DeanWay** [comment]
@sdumitriu test coverage is now %100

**Pull request merged**